### PR TITLE
Implement UserStore

### PR DESCRIPTION
**Pull request by tomrf1**
Update and retrieve users from dynamodb.

I've added a `version` field to prevent data being overwritten when 2 messages from the same user are being processed concurrently. This effectively does the same thing as [optimistic locking](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.OptimisticLocking.html), but that uses java annotations and probably won't work with scanamo, plus it's extremely easy to implement here using conditional updates.

I still need to add [read consistency](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html), but I first need to add support to scanamo.

### CODE REVIEW

File: src/main/scala/com/gu/facebook_news_bot/BotService.scala [hunk 1]
```diff
@@ -26,19 +27,19 @@ trait BotService extends CirceSupport {
...
-        new AmazonDynamoDBClient().withEndpoint("http://localhost:8000")
+        new AmazonDynamoDBAsyncClient().withEndpoint("http://localhost:8000")
       } else {
         val awsRegion = Regions.fromName(BotConfig.aws.region)
-        new AmazonDynamoDBClient(BotConfig.aws.CredentialsProvider).withRegion(awsRegion)
+        new AmazonDynamoDBAsyncClient(BotConfig.aws.CredentialsProvider).withRegion(awsRegion)
       }
     }
-    new UserStore(new DynamoDB(client), usersTable)
+    new UserStore(client, usersTable, logger)
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/facebook_news_bot/BotService.scala]: Can you use the Typesafe logger which provides the `StrictLogging` trait that can be extended so you don't have to pass logger around?

File: src/main/scala/com/gu/facebook_news_bot/models/User.scala [hunk 2]
```diff
@@ -3,9 +3,10 @@ package com.gu.facebook_news_bot.models
 /**
   * The user data as it is stored in dynamodb
   */
-case class User(id: String,
+case class User(ID: String,
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/facebook_news_bot/models/User.scala]: Why is this field capitalised?

File: src/main/scala/com/gu/facebook_news_bot/models/FacebookModels.scala [hunk 3]
```diff
@@ -13,8 +13,14 @@ object MessageToFacebook {
 
   def textMessage(id: String, message: String) =
     MessageToFacebook(Id(id),
-               Some(Message(Some(message),None,None,None)),
-               None, None)
+     Some(Message(Some(message),None,None,None)),
+     None, None)
+
+  def errorMessage(id: String) =
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/facebook_news_bot/models/FacebookModels.scala]: Can we better model the different types of messages as an ADT?

File: src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
+package com.gu.facebook_news_bot
```
Comment (LATaylor-guardian) [inline src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala]: Should this package be `com.gu.facebook-news-bot`. Seems unusual to use underscores in a package name in a Scala project.

File: src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+import akka.actor.ActorSystem
+import akka.event.Logging
+import com.amazonaws.services.dynamodbv2.model.ScalarAttributeType._
+import com.gu.facebook_news_bot.models.User
+import com.gu.facebook_news_bot.stores.UserStore
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.time.{Millis, Seconds, Span}
+
+class UserStoreTest extends org.scalatest.FunSpec with org.scalatest.Matchers with ScalaFutures {
```
Comment (LATaylor-guardian) [inline src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala]: import import org.scalatest.{FunSpec, Matchers}

File: src/main/scala/com/gu/facebook_news_bot/stores/UserStore.scala [hunk 6]
```diff
@@ -1,21 +1,45 @@
...
-  private val usersTable = db.getTable(usersTableName)
+/**
+  * TODO -
+  * Make reads strongly consistent - requires scanamo change
+  */
+class UserStore(client: AmazonDynamoDBAsyncClient, usersTableName: String, logger: LoggingAdapter) {
 
-  def getUser(id: String): Option[User] = {
-    None
+  def getUser(id: String): Future[Option[User]] = {
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/facebook_news_bot/stores/UserStore.scala]: All public functions should have return types.

File: src/main/scala/com/gu/facebook_news_bot/models/User.scala (hunk 2 above)
Comment (tomrf1) [inline src/main/scala/com/gu/facebook_news_bot/models/User.scala]: To be consistent with the field we currently have in dynamodb, not sure if it matters

File: src/main/scala/com/gu/facebook_news_bot/models/FacebookModels.scala (hunk 3 above)
Comment (tomrf1) [inline src/main/scala/com/gu/facebook_news_bot/models/FacebookModels.scala]: This is the only error message I was planning on sending to the user, the actual cause is logged elsewhere.

File: src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala (hunk 4 above)
Comment (tomrf1) [inline src/test/scala/com/gu/facebook_news_bot/UserStoreTest.scala]: I wish, but java (and therefore scala) doesn't allow dashes in package names. The [suggested](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html) alternative is undescores

File: src/main/scala/com/gu/facebook_news_bot/models/User.scala (hunk 2 above)
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/facebook_news_bot/models/User.scala]: I'd revert to id if it doesn't break anything.

Comment (tomrf1) [inline src/main/scala/com/gu/facebook_news_bot/models/User.scala]: Just tested this - the case does need to be consistent, so I'll leave it capitalised

**Pull request merged**